Perf percentage math

MerlinEgalite opened the pull request:
We could improve the average computation if we do the following:
`(x * (1 - percentage) + y * percentage) / PERCENTAGE_FACTOR` instead of:
`x * (1 - percentage)  / PERCENTAGE_FACTOR + y * percentage / PERCENTAGE_FACTOR`

Note that it's not the same computation with the roundings and also it would overflow with a lower (x, y) pair.

MerlinEgalite committed:
⚡️ Improve percentage add

MerlinEgalite committed:
🩹 Fix check in percentage add

MerlinEgalite committed:
⚡️ Optimize add further

MerlinEgalite committed:
⚡️ Do the same for sub

MerlinEgalite committed:
🎨 Improve readability

MerlinEgalite committed:
⚡️ Optimize percentage average

MerlinEgalite committed:
🎨 Lint code

MerlinEgalite committed:
⚡️ Optimize further

MerlinEgalite committed:
✏️ Fix wtf typo

Rubilmax commented:
> We could improve the average computation if we do the following:
> `(x * (1 - percentage) + y * percentage) / PERCENTAGE_FACTOR` instead of:
> `x * (1 - percentage)  / PERCENTAGE_FACTOR + y * percentage / PERCENTAGE_FACTOR`
> 
> Note that it's not the same computation with the roundings and also it would overflow with a lower (x, y) pair.

Definitely to avoid because of the 2 points you made

Will check this PR out, thanks!

MathisGD commented on src/math/PercentageMath.sol:
```diff
@@ -72,12 +104,41 @@ library PercentageMath {
+        //
+        // Overflow if x2 = y * percentage + HALF_PERCENTAGE_FACTOR > type(uint256).max
+        // <=> y * percentage > type(uint256).max - HALF_PERCENTAGE_FACTOR
+        // <=> y > (type(uint256).max - HALF_PERCENTAGE_FACTOR) / percentage
```
percentage should be greater than 0

MerlinEgalite replied:
let percentage > 0

MathisGD replied:
yes, but it is weird this way IMO

MathisGD reviewed (commented):
I'm opening a PR on top of this one to fix all of this

MerlinEgalite commented on src/math/PercentageMath.sol:
```diff
@@ -18,17 +18,48 @@ library PercentageMath {
+                y,
+                or(
+                    gt(percentage, sub(MAX_UINT256, PERCENTAGE_FACTOR)),
+                    gt(x, div(MAX_UINT256_MINUS_HALF_PERCENTAGE, y))
```
wondering the same thing, we know that operations between constants is optimized after compilation, so either we only use the operations or the long constants even we know that it does not bring any gas optimizations.

MerlinEgalite commented on src/math/PercentageMath.sol:
```diff
@@ -18,17 +18,48 @@ library PercentageMath {
+            y := add(percentage, PERCENTAGE_FACTOR) // Temporary assignement to save gas.
+
+            if mul(
+                y,
```
not sure to understand sorry

MathisGD replied:
percentage <= MAX - PERCENTAGE_FACTOR => percentage + PERCENTAGE_FACTOR < MAX + 1 => y != 0

so no need to verify that y != 0

MathisGD commented on src/math/PercentageMath.sol:
```diff
@@ -72,12 +104,41 @@ library PercentageMath {
+                    ),
+                    mul(percentage, gt(y, div(MAX_UINT256_MINUS_HALF_PERCENTAGE, percentage)))
+                ),
+                lt(z, x1)
```
x1 + x2 is always less than MAX
knowing this, we can optimize a lot this one

MerlinEgalite replied:
hmm why?

MathisGD replied:
Because x1 + x2 <= max(x, y)

MathisGD committed:
doc: improve natspec doc

MathisGD committed:
perf: opti percentAdd

y = PERCENTAGE_FACTOR + percentage > 0 so we don't need to verify that y is greater than 0 before dividing by y.

PERCENTAGE_FACTOR + percentage instead of percentage + PERCENTAGE_FACTOR because I find it more natural.

MathisGD committed:
perf: opti percentSub

AND is cheaper than MUL, and putting it after might allow to skip it if the first assertion of the OR is true.

MathisGD committed:
doc: improve percentMul and percentDiv doc

MathisGD committed:
perf: opti weightedAvg

(x * (PERCENTAGE_FACTOR - percentage) + HALF_PERCENTAGE_FACTOR / PERCENTAGE_FACTOR) + (y * percentage + HALF_PERCENTAGE_FACTOR / PERCENTAGE_FACTOR) <= max(x, y) < MAX

so we don't need to verify that the last addition does not overflow.

MathisGD commented on src/math/PercentageMath.sol:
```diff
@@ -72,12 +104,41 @@ library PercentageMath {
+        //
+        // Overflow if x1 = x * (PERCENTAGE_FACTOR - percentage) + HALF_PERCENTAGE_FACTOR > type(uint256).max
+        // <=> x * (PERCENTAGE_FACTOR - percentage) > type(uint256).max - HALF_PERCENTAGE_FACTOR
+        // <=> x > (type(uint256).max - HALF_PERCENTAGE_FACTOR) / (PERCENTAGE_FACTOR - percentage)
```
`(PERCENTAGE_FACTOR - percentage)` should be greater than 0

MerlinEgalite replied:
first condition if not passed then it reverts

MathisGD replied:
Not exactly 
There is the case percentage = PERCENTAGE_FACTOR

MerlinEgalite replied:
Oh ok understand now what you mean

MathisGD committed:
test: improve PercentageMath tests

MathisGD committed:
doc: minor improvement

MathisGD committed:
fix: silent overflow and fuzz this case

MathisGD committed:
test: clean test by using mathRef

MathisGD committed:
doc: minor fix

MathisGD committed:
test: use expectRevert instead of testFail

MathisGD committed:
doc: improve doc

MathisGD committed:
doc: minor fix

Co-authored-by: Quentin Garchery <QGarchery@users.noreply.github.com>

MathisGD committed:
fix: weighted average double round up

MathisGD committed:
perf: delete a redondant overflow condition

MathisGD committed:
test: use constants

MathisGD committed:
doc: fix coments weightedAvg

MathisGD committed:
perf: reorder subs

MathisGD committed:
Merge pull request #43 from morpho-dao/fix/weightedAvg

Fix weighted average double round up

MathisGD committed:
Merge pull request #41 from morpho-dao/perf/percentage-math-1

Fixes proposal after review
